---
title: Introduction to mvsusie for multi-trait fine-mapping
author: Yuxin Zou, Gao Wang, Peter Carbonetto and Matthew Stephens
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to mvsusie for multi-trait fine-mapping}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

The aim of this vignette is to work through the steps of a basic
mvSuSiE multi-trait fine-mapping analysis. We analyze a simulated
fine-mapping data set with realistic genotypes.

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,results = "hold",
                      fig.align = "center",dpi = 120)
```

First, load the necessary packages, and to ensure reproducibility set
the seed.

```{r load-pkgs, message=FALSE}
library(mvsusieR)
library(ggplot2)
library(cowplot)
set.seed(1)
```

Next, load the data set we will analyze.

```{r load-data}
data(simdata)
```

The data consist of an $N \times J$ genotype matrix `X` and an $N
\times R$ phenotype matrix `Y` where $N$ is the number of samples, $J$
is the number of genetic markers, and $R$ is the number of
traits/phenotypes.

```{r inspect-data}
X <- simdata$raw$X
Y <- simdata$raw$Y
dim(X)
dim(Y)
```

Specify the prior
-----------------

Some care needs to be taken to choose the prior for your mvSuSiE
analysis, particularly when you are analyzing many traits. In this
example, we have defined the prior for you:

```{r create-prior}
prior <- create_mixture_prior(list(matrices = simdata$par$U,
                                   weights = simdata$par$w),
                              null_weight = 0)
```
							  
More generally, we recommend taking a "data-driven" approach to
estimating the weights and the covariance matrices in the prior. For
example, when there are many candidate fine-mapping regions in your
study you could fit the prior to all the top associations from each of
the candidate fine-mapping regions. This approach is quite similar to
the approach used in [mashr][mashr], and we recommend looking through
the vignettes in the mashr package for more detailed guidance.

Fit the mvSuSiE model
---------------------

Now that we have set up the prior, we are ready to run mvsusie:

```{r run-mvsusie}
fit <- mvsusie(X,Y,standardize = TRUE,
               prior_variance = prior,
               residual_variance = simdata$par$V,
			   estimate_prior_variance = TRUE,
			   tol = 0.01)
```

In this call, we have provided mvSuSiE with the four key pieces of
information need to run the fine-mapping analysis: (1) the genotype
matrix; (2) the phenotype matrix, (3) the prior on the effects; and
(4) the residual variance-covariance matrix (an $R \times R$ matrix). 

By default, `mvsusie` fits a model with L = 10 single effects. mvSuSiE
is generally robust to misspecification of L so long as L is chosen to
be larger than the true number of effects. mvSuSiE prunes unneeded
single effects by estimating a scaling coefficient in the prior; we
set `estimate_prior_variance = TRUE` to activate this step. Indeed,
`mvsusie` prunes 7 out of 10 of the single effects, resulting in 3
(95%) credible sets (CSs), each of which contain exactly one of the
true causal variants:

```{r credible-sets}
fit$sets$cs
```

In the third CS, mvSuSiE cannot decide which genetic variant is the
causal variant because the candidate variants are strongly correlated
with each other:

```{r cs-purity}
fit$sets$purity
```

Examining the posterior inclusion probabilties (PIPs) for these
variants, there is a subset of three variants that account for >58% of
probability for being causal:

```{r cs-3}
markers <- fit$sets$cs$L3
fit$pip[markers]
```

Additionally, the "sentinel" variant---the variant with the highest
PIP---as a PIP of nearly 40%.

In a multi-trait analysis, we are not just interested in identifying
the causal SNPs, but also _which traits_ are affected by these causal
SNPs.

```{r, fig.height=2.5, fig.width=3.5}
l <- 1
i <- 335
pdat <- data.frame(bhat = simdata$Btrue[i,],
                   coef = fit$b1_rescaled[l,i+1,],
                   lfsr = fit$single_effect_lfsr[l,])
ggplot(pdat,aes(x = bhat,y = coef,color = log10(lfsr),pch = 20)) +
  geom_point(shape = 20,size = 2.5) +
  geom_abline(intercept = 0,slope = 1,color = "black",linetype = "dotted") +
  scale_color_gradient2(low = "lightskyblue",mid = "gold",high = "orangered",
                        midpoint = -40) +
  labs(x = "true coef",y = "mvsusie estimate") +
  theme_cowplot(font_size = 12)
```

```{r, fig.height=2.5, fig.width=3.5}
l <- 3
i <- 493
pdat <- data.frame(bhat = simdata$Btrue[i,],
                   coef = fit$b1_rescaled[l,i+1,],
                   lfsr = fit$single_effect_lfsr[l,])
ggplot(pdat,aes(x = bhat,y = coef,color = log10(lfsr),pch = 20)) +
  geom_point(shape = 20,size = 2.5) +
  scale_color_gradient2(low = "lightskyblue",mid = "gold",high = "orangered",
                        midpoint = -3) +
  labs(x = "true coef",y = "mvsusie estimate") +
  theme_cowplot(font_size = 12)
```

mvSuSiE with summary data: mvSuSiE-RSS
--------------------------------------

```{r mvsusie-rss}
n <- simdata$sumstats$n
Z <- simdata$sumstats$z
R <- simdata$sumstats$LD
fit_rss <- mvsusie_rss(Z,R,n,
                       prior_variance = prior,
					   residual_variance = simdata$par$V,
					   estimate_prior_variance = TRUE,
					   tol = 0.01)
```

```{r}
l <- 1
i <- 335
fit$b1[l,i,]
fit_rss$b1[l,i,]
```

[mashr]: https://github.com/stephenslab/mashr
